#!/bin/bash
#SBATCH -A PAUL-SL3-CPU
#SBATCH -p skylake-himem
#SBATCH --mem 50G
#SBATCH --job-name=filter_chrx
#SBATCH --time=12:0:0
#SBATCH --output=/rds/project/jmmh2/rds-jmmh2-projects/interval_rna_seq/covid19/genotypes/plink_filter_log_%A_%a.log
#SBATCH --mail-type=ALL
#SBATCH --mail-user=jm2294@medschl.cam.ac.uk

. /etc/profile.d/modules.sh     
module purge                  
module load rhel7/default-peta4
module load plink-1.9-gcc-5.4.0-sm3ojoi

cd /rds/project/jmmh2/rds-jmmh2-projects/interval_rna_seq/covid19/genotypes

plink\
 --vcf ~/rds/rds-jmmh2-projects/covid/ace2/interval_genetic_data/interval_imputed_data/INTERVAL_X_imp_ann_filt_v2.vcf.gz\
 --chr 23\
 --keep /rds/project/jmmh2/rds-jmmh2-projects/interval_rna_seq/analysis/04_phase2_full_analysis/genotypes/processing/rna_seq_phase1-2_affy_ids.txt\
 --make-bed\
 --out INTERVAL_chrX_merged_cleaned_RNAseq_phase1-2
 
module load plink
# Generate ids for SNPs without RSid and filter on MAF 
plink2\
 --bfile INTERVAL_chrX_merged_cleaned_RNAseq_phase1-2\
 --chr 23\
 --output-chr 26\
 --set-missing-var-ids @:#_\$r_\$a\
 --rm-dup exclude-mismatch\
 --new-id-max-allele-len 20 truncate\
 --make-bed\
 --maf 0.005\
 --geno 0.05\
 --out INTERVAL_chrX_merged_cleaned_RNAseq_phase1-2_deduplicated_MAF0.005
 
 # Filter ACE2 ids only, requires file generated by cov19_make_bed.r
 plink2\
 --bfile INTERVAL_chrX_merged_cleaned_RNAseq_phase1-2_deduplicated_MAF0.005\
 --chr 23\
 --output-chr 26\
 --keep ace2_nonzero_ids.txt\
 --make-bed\
 --maf 0.005\
 --out INTERVAL_chrX_merged_cleaned_RNAseq_phase1-2_deduplicated_MAF0.005_ACE2nonzero
